src/spetlr/power_bi/PowerBi.py

```diff
@@ -0,0 +1,459 @@
+            if no parameter was specified.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if self.dataset_id is None:
+            api_call = requests.get(
+                url=f"{self.powerbi_url}groups/{self.workspace_id}/datasets",
+                headers=self.api_header,
+            )
+            if api_call.status_code != 200:
+                self._raise_api_error("Failed to fetch datasets!", api_call)
+                return False
+            df = pd.DataFrame(api_call.json()["value"], columns=["id", "name"])
+
+            if self.dataset_name is None:
+                self._show_datasets(df)
+                return False
+
+            rows = df.loc[df["name"] == self.dataset_name, "id"]
+            if rows.empty:
+                self._raise_error(
+                    f"Dataset name '{self.dataset_name}' cannot be found, "
+                    "or the dataset doesn't have a user with the required permissions!"
+                )
+                return False
+            self.dataset_id = rows.values[0]
+
+        return True
+
+    def _connect(self) -> bool:
+        """
+        Connects or reconnects to PowerBI to fetch refresh history or trigger a refresh.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if not self._get_access_token():
+            return False
+        if not self._get_workspace():
+            return False
+        if not self._get_dataset():
+            return False
+
+        return True
+
+    def _get_last_refresh(self) -> bool:
+        """
+        Gets the latest record in the PowerBI dataset refresh history.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if not self._connect():
+            return False
+
+        api_url = (
+            f"{self.powerbi_url}groups/{self.workspace_id}"
+            f"/datasets/{self.dataset_id}/refreshes?$top=1"
+        )
+        api_call = requests.get(url=api_url, headers=self.api_header)
+        if api_call.status_code == 200:
+            json = api_call.json()
+            df = pd.DataFrame(
+                json["value"],
+                columns=[
+                    "requestId",
+                    "id",
+                    "refreshType",
+                    "startTime",
+                    "endTime",
+                    "status",
+                    "serviceExceptionJson",
+                ],
+            )
+            df.set_index("id")
+            if len(df.index) > 0:
+                self.last_status = df.status[0]
+                self.last_exception = df.serviceExceptionJson[0]
```

Comment:
I guess this is subjetive, but consider declare these variables:
_endtime = df.endTime[0]
_starttime=df.startTime[0]
In this way, it is only defined once. Then replace them!

Reply:
Agreed. Implemented here:
https://github.com/spetlr-org/spetlr/pull/149